src/absil/illib.fs

```diff
@@ -959,16 +960,18 @@ let _ = eventually { use x = null in return 1 }
 
 /// Generates unique stamps
 type UniqueStampGenerator<'T when 'T : equality>() = 
-    let encodeTab = new Dictionary<'T, int>(HashIdentity.Structural)
+    let gate = obj ()
+    let encodeTab = new ConcurrentDictionary<'T, int>(HashIdentity.Structural)
     let mutable nItems = 0
     let encode str =
         match encodeTab.TryGetValue str with
         | true, idx -> idx
         | _ ->
-            let idx = nItems
-            encodeTab.[str] <- idx
-            nItems <- nItems + 1
-            idx
+            lock gate (fun () ->
+                let idx = nItems
+                encodeTab.[str] <- idx
+                nItems <- nItems + 1
+                idx)
```

Comment:
You can do this lock-free with something like:
```suggestion
            let idx = Interlocked.Increment &nItems
            encodeTab.[str] <- idx
            idx
```

(and initializing `nItems = -1` above so the first value returned by Interlocked.Increment is 0)

Reply:
The lock isn't just to do the increment. The lock is to prevent a key, `str`, from being associated with potentially different values. Meaning this could happen at the same time:
```encodeTab.["test"] <- 0```
```encodeTab.["test"] <- 1```

I don't know a better way to handle it, but I think it's ok.